### PR TITLE
Fix for link restrictions

### PR DESCRIPTION
**Pull request by Gulfa**

### CODE REVIEW

File: meerkat_abacus/data_management.py [hunk 1]
```diff
@@ -873,7 +886,7 @@ def new_data_to_codes(engine=None, debug_enabled=True, restrict_uuids=None,
     country_config = param_config.country_config
 
     if restrict_uuids is not None:
-        if restrict_uuids == []:
+        if restrict_uuids == {} or sum([len(v) for v in restrict_uuids.values()]) == 0:
```
Comment (tomeksabala) [inline meerkat_abacus/data_management.py]: Could you give this second condition `sum([len(v) for v in restrict_uuids.values()]) == 0` a name in form of local variable?

File: meerkat_abacus/data_management.py [hunk 2]
```diff
@@ -895,10 +908,16 @@ def new_data_to_codes(engine=None, debug_enabled=True, restrict_uuids=None,
...
+        if data_type["form"] not in restrict_uuids:
+            link_forms = []
+            for link in links_by_type.get(data_type["name"], []):
+                link_forms += [link["to_form"], link["from_form"]]
+            found = False
+            for l in set(link_forms):
+                if l in restrict_uuids.keys():
+                    found = True
+            if found is False:
+                continue
```
Comment (tomeksabala) [inline meerkat_abacus/data_management.py]: Why do you create a set from link_forms?
`if found is False:` looks evil. I'd go with `if not found:` or even better sth like this:
```python
if any([link not in restrict_uuids.keys() for link in set(link_forms)]):
  continue
```

File: meerkat_abacus/data_management.py [hunk 3]
```diff
@@ -924,9 +943,12 @@ def new_data_to_codes(engine=None, debug_enabled=True, restrict_uuids=None,
                                    engine=engine)
         # Main Query
         if restrict_uuids is not None:
+            all_uuids = []
+            for form_uuids in restrict_uuids.values():
+                all_uuids += form_uuids
```
Comment (tomeksabala) [inline meerkat_abacus/data_management.py]: I wonder if such a thing would work.
`all_uuids = list(itertools.chain.from_iterable(restrict_uuids.values()))`
https://docs.python.org/3/library/itertools.html#itertools.chain.from_iterable

File: meerkat_abacus/data_management.py [hunk 4]
```diff
@@ -976,6 +999,7 @@ def new_data_to_codes(engine=None, debug_enabled=True, restrict_uuids=None,
             if debug_enabled:
                 logging.debug("Added %s records", added)
         if data:
+            #logging.info(data)
```
Comment (tomeksabala) [inline meerkat_abacus/data_management.py]: Stale debug statement?

File: meerkat_abacus/data_management.py [hunk 5]
```diff
@@ -962,6 +984,7 @@ def new_data_to_codes(engine=None, debug_enabled=True, restrict_uuids=None,
                 # Send all data apart from the latest UUID to to_data function
                 last_data = data.pop(uuid)
                 if data:
+
```
Comment (tomeksabala) [inline meerkat_abacus/data_management.py]: Extra new line.

File: meerkat_abacus/pipeline.py [hunk 6]
```diff
@@ -65,7 +65,7 @@ def process_chunk(internal_buffer, session, engine, param_config=config,
 
     """
     start = time.time()
-    uuids = []
+    uuids = {}
```
Comment (tomeksabala) [inline meerkat_abacus/pipeline.py]: I think that we should rename this dict to `uuids_to_forms_map` or sth.

File: meerkat_abacus/pipeline.py [hunk 7]
```diff
@@ -81,15 +81,19 @@ def process_chunk(internal_buffer, session, engine, param_config=config,
             session,
             engine,
             **kwargs)
-        uuids += new_uuids
+        uuids.setdefault(form, [])
```
Comment (tomeksabala) [inline meerkat_abacus/pipeline.py]: uuids could be a default dict.
https://stackoverflow.com/a/3483652
```python
uuids = defaultdict(list)
```
Also the above post suggests a little more efficient and readable way:
```python
form_uuids = uuids.setdefault(form,[])
form_uuids.append(new_uuids)
```

File: meerkat_abacus/pipeline.py [hunk 8]
```diff
@@ -81,15 +81,19 @@ def process_chunk(internal_buffer, session, engine, param_config=config,
...
         if len(new_uuids) > 0:
             forms.append(form)
     corrected = data_management.initial_visit_control(
         param_config=param_config
     )
-    uuids += corrected
+    corrected_tables = list(
+        param_config.country_config['initial_visit_control'].keys())
+    if corrected_tables:
+        if corrected_tables[0] in uuids:
```
Comment (tomeksabala) [inline meerkat_abacus/pipeline.py]: Dictionaries are not ordered in python. Why are we checking the first key here?

File: meerkat_abacus/data_management.py (hunk 2 above)
Comment (Gulfa) [reply meerkat_abacus/data_management.py]: just creating a set to remove any duplicates in the list

File: meerkat_abacus/data_management.py (hunk 3 above)
Comment (Gulfa) [reply meerkat_abacus/data_management.py]: Probably would work, but I am not sure it is that much clearer.

Comment (tomeksabala) [reply meerkat_abacus/data_management.py]: Ok. It's not but I was sure it'd be faster.
But I've tested the performance and if we need a `list` type to pass later to `Column.in_` function then the original approach is almost twice faster.

File: meerkat_abacus/pipeline.py [hunk 9]
```diff
@@ -5,7 +5,7 @@
 import logging
 from collections import defaultdict
 from meerkat_abacus import data_import
-from meerkat_abacus import util
+from meerkat_abacus import util{{ _(
```
Comment (tomeksabala) [inline meerkat_abacus/pipeline.py]: Typo?

Comment (Gulfa) [reply meerkat_abacus/pipeline.py]: Yes

**Pull request merged**